llvm/lib/Transforms/Yk/ShadowStack.cpp

```diff
@@ -0,0 +1,230 @@
+            size_t Align = AI.getAlign().value();
+            Offset = int((Offset + (Align - 1)) / Align) * Align;
+            GetElementPtrInst *GEP = GetElementPtrInst::Create(Int8Ty, SSPtr, {ConstantInt::get(Int32Ty, Offset)}, "", cast<Instruction>(&AI));
+            Builder.SetInsertPoint(GEP);
+            Builder.CreateBitCast(GEP, AI.getAllocatedType()->getPointerTo());
+            cast<Value>(I).replaceAllUsesWith(GEP);
+            RemoveAllocas.push_back(cast<Instruction>(&AI));
+            Offset += AllocaSize;
+          } else if (isa<CallInst>(I)) {
+            // When we see a call, we need make space for a new stack frame. We
+            // do this by simply adjusting the pointer stored in the global
+            // shadow stack. When the function returns the global is reset. This
+            // is similar to how the RSP is adjusted inside the
+            // prologue/epilogue of a function, but here the prologue/epilogue
+            // are handled by the caller.
+            CallInst &CI = cast<CallInst>(I);
+            if (&CI == Malloc) {
+              // Don't do this for `malloc` that created the shadow stack.
+              continue;
+            }
+            // Inline asm can't be traced.
+            if (CI.isInlineAsm()) {
+              continue;
+            }
+
+            // YKFIXME: Skip functions that are marked with `yk_outline`
+            // (as those won't be traced and thus don't require a shadow
+            // stack).
+            // YKFIXME: Skip functions (direct or indirect) that we don't have
+            // IR for.
+
+            if (CI.getCalledFunction()) {
+              if (CI.getCalledFunction()->getName() == "pthread_create") {
+                // The global shadow stack needs to be thread local. In each new
+                // thread created we need to malloc a new shadow stack and
+                // assign it to that threads shadow stack global. Note: it's not
+                // enough to look for `pthread_create` in here, as this call
+                // could be hidden inside an external library.
+                Context.emitError("Unable to add shadow stack: No support for threads yet!");
+                return false;
+              }
+              // Skip some known intrinsics. YKFIXME: Is there a more general solution, e.g. skip all intrinsics?
+              else if (CI.getCalledFunction()->getName() == "llvm.experimental.stackmap") {
+                continue;
+              }
+              else if (CI.getCalledFunction()->getName() == "llvm.dbg.declare") {
+                continue;
+              }
+            }
+
+            // Adjust shadow stack pointer before a call, and reset it back to
+            // its previous value upon returning. Make sure to align the shadow
+            // stack to a 16 byte boundary before calling, as required by the
+            // calling convention.
+            Offset = int((Offset + (16 - 1)) / 16) * 16;
```

Comment:
This is X86 specific so needs `#ifdef`ing.

Reply:
Fixed in [371f02b](https://github.com/ykjit/ykllvm/pull/53/commits/371f02b441565526a6bb46d559e2fd0fb811ebef)